### PR TITLE
fix: Use DEFAULT_REQUEUE_TIME for Agent. Fixes #7269

### LINKED ISSUE #7269: Agent patches task-set too often
Every 1s in fact. But that is very wasteful. It only needs to patch once per DEFAULT_REQUEUE_TIME, 10s by default. So reduce to every 10s.

### PR DESCRIPTION
**Pull request by simster7**
Fixes https://github.com/argoproj/argo-workflows/issues/7269

Signed-off-by: Simon Behar <simbeh7@gmail.com>

Don't bother creating a PR until you've done this:

* [ ] Run `make pre-commit -B` to fix codegen, lint, and commit message problems.

Create your PR as a draft.

* Your PR needs to pass the required checks before it can be approved. If the check is not required (e.g. E2E tests) it
  does not need to pass.
* Once required tests have passed, you can make it "Ready for review".
* Say how how you tested your changes. If you changed the UI, attach screenshots.

Tips:

* If changes were requested, and you've made them, then dismiss the review to get it looked at again.
* Add you organization to [USERS.md](https://github.com/argoproj/argo-workflows/blob/master/USERS.md) if you like.
* You can ask for help!

### CODE REVIEW

File: workflow/common/common.go [hunk 1]
```diff
@@ -132,6 +131,8 @@ const (
 	EnvVarProgressFileTickDuration = "ARGO_PROGRESS_FILE_TICK_DURATION"
 	// EnvVarProgressFile is the file watched for reporting progress
 	EnvVarProgressFile = "ARGO_PROGRESS_FILE"
+	// EnvVarDefaultRequeueTime is the default requeue time for Workflow Informers. For more info, see rate_limiters.go
+	EnvVarDefaultRequeueTime = "DEFAULT_REQUEUE_TIME"
```
Comment (alexec) [inline workflow/common/common.go:135]: nice, only request `ARGO_DEFAULT_REQUEUE_TIME`

Comment (simster7) [reply workflow/common/common.go:135]: @alexec I have not created a new variable here, but simply reused the existing one. My fear is that changing the variable name will break users who have `DEFAULT_REQUEUE_TIME` hard wired. I can always accept both if necessary?

Comment (alexec) [reply workflow/common/common.go:135]: ah - prior art - ok

File: workflow/controller/rate_limiters.go [hunk 2]
```diff
@@ -6,19 +6,24 @@ import (
...
+func GetRequeueTime() time.Duration {
 	// We need to rate limit a minimum 1s, otherwise informers are unlikely to be upto date
 	// and we'll operate on an out of date version of a workflow.
 	// Under high load, the informer can get many seconds behind. Increasing this to 30s
 	// would be sensible for some users.
 	// Higher values mean that workflows with many short running (<20s) nodes do not progress as quickly.
 	// So some users may wish to have this as low as 2s.
 	// The default of 10s provides a balance more most users.
-	return env.LookupEnvDurationOr("DEFAULT_REQUEUE_TIME", 10*time.Second)
+	return env.LookupEnvDurationOr(common.EnvVarDefaultRequeueTime, 10*time.Second)
```
Comment (alexec) [inline workflow/controller/rate_limiters.go:20]: could we make this a global, and log on start-up?

File: workflow/executor/agent.go [hunk 3]
```diff
@@ -63,13 +63,14 @@ func (ae *AgentExecutor) Agent(ctx context.Context) error {
 	defer runtimeutil.HandleCrash(runtimeutil.PanicHandlers...)
 
 	taskWorkers := env.LookupEnvIntOr(EnvAgentTaskWorkers, 16)
-	log.WithField("task_workers", taskWorkers).Info("Starting Agent s15")
+	requeueTime := env.LookupEnvDurationOr(common.EnvVarDefaultRequeueTime, 10*time.Second)
+	log.WithFields(log.Fields{"task_workers": taskWorkers, "requeue_time": requeueTime}).Info("Starting Agent a3")
```
Comment (alexec) [inline workflow/executor/agent.go:67]: what is `a3`?

Comment (alexec) [reply workflow/executor/agent.go:67]: minor - I think we use camel case, e.g. `taskWorkers` not `task_workers`

File: workflow/common/common.go (hunk 1 above)
Comment (alexec) [reply workflow/common/common.go:135]: thought - do we want to make this independently configurable? that way, any problems, we can change them independently

Comment (simster7) [reply workflow/common/common.go:135]: Sorry, not sure what you mean by independently configurable?

File: workflow/executor/agent.go (hunk 3 above)
Comment (simster7) [reply workflow/executor/agent.go]: Fixed

File: workflow/controller/rate_limiters.go (hunk 2 above)
Comment (simster7) [reply workflow/controller/rate_limiters.go:20]: I added logging on start-up. This function is globally available, we can make it a variable itself, but I think I would prefer the function wrapper. Thoughts?

File: workflow/common/common.go (hunk 1 above)
Comment (alexec) [reply workflow/common/common.go:135]: I think we should be able to cater for the agent going crazy and we want to reduce its patch periodicity, but not the controller's re-queue time. That would mean two configuration items.

Comment (simster7) [reply workflow/common/common.go:135]: Sure thing, I can allow for two configuration items. If no configuration is found specifically for agent patch periodicity, I will fall back to controller's re-queue time as default.

File: workflow/controller/rate_limiters.go (hunk 2 above)
Comment (alexec) [reply workflow/controller/rate_limiters.go:20]: no strong opinion

**Comment by simster7** [comment]
@alexec Done, re-review please

File: workflow/controller/agent.go [hunk 4]
```diff
@@ -82,9 +82,14 @@ func (woc *wfOperationCtx) createAgentPod(ctx context.Context) (*apiv1.Pod, erro
 		}
 	}
 
+	agentPatchRate := GetRequeueTime().String()
```
Comment (alexec) [inline workflow/controller/agent.go:85]: minor - encapsulate into a func you can then test

File: workflow/executor/agent.go [hunk 5]
```diff
@@ -57,13 +57,16 @@ type response struct {
 	Result *wfv1.NodeResult
 }
 
-const EnvAgentTaskWorkers = "ARGO_AGENT_TASK_WORKERS"
+const (
+	EnvAgentTaskWorkers = "ARGO_AGENT_TASK_WORKERS"
```
Comment (alexec) [inline workflow/executor/agent.go:61]: can we put these env vars in common.go?

File: workflow/controller/controller.go [hunk 6]
```diff
@@ -223,8 +223,8 @@ func (wfc *WorkflowController) Run(ctx context.Context, wfWorkers, workflowTTLWo
 	defer wfc.podQueue.ShutDown()
 	defer wfc.podCleanupQueue.ShutDown()
 
-	log.WithField("version", argo.GetVersion().Version).Info("Starting Workflow Controller")
-	log.Infof("Workers: workflow: %d, pod: %d, pod cleanup: %d", wfWorkers, podWorkers, podCleanupWorkers)
+	log.WithFields(log.Fields{"version": argo.GetVersion().Version, "defaultRequeueTime": GetRequeueTime()}).Info("Starting Workflow Controller")
+	log.WithFields(log.Fields{"workflow": wfWorkers, "workflowTtl": workflowTTLWorkers, "pod": podWorkers, "podCleanup": podCleanupWorkers}).Infof("Current Worker Numbers")
```
Comment (alexec) [inline workflow/controller/controller.go:227]: I like to use train-wreck pattern for this now as I think it is easier to read:

```golnga
log.WithField("wfWorkers", wfWorkers).
  WithField("workflowTTLWorkers", workflowTTLWorkers).
  Info("this is the config")
```

**Review by alexec — CHANGES_REQUESTED** [review]
small changes

File: workflow/controller/agent.go (hunk 4 above)
Comment (simster7) [reply workflow/controller/agent.go]: Done

File: workflow/executor/agent.go (hunk 5 above)
Comment (simster7) [reply workflow/executor/agent.go]: Done

File: workflow/controller/controller.go (hunk 6 above)
Comment (simster7) [reply workflow/controller/controller.go]: Done

**Comment by simster7** [comment]
@alexec All done

File: util/env/env.go [hunk 7]
```diff
@@ -46,3 +46,11 @@ func LookupEnvFloatOr(key string, o float64) float64 {
 	}
 	return o
 }
+
+func LookupEnvStringOr(key string, o string) string {
```
Comment (alexec) [inline util/env/env.go:50]: I think this might already exist somewhere else?

**Review by alexec — APPROVED** [review]

**Pull request merged**